### ISSUE TITLE
osu occasionally crashes WindowServer, causing entire Mac to freeze and eventually shut down

### ISSUE THREAD

**Issue by chaNcharge** [ppy/osu#8278]
**Describe the crash:** 
It appears to happen seemingly randomly, usually quicker on a fresh boot, WindowServer crashes, which ends up freezing the entire computer, forcing me to either force shut down or wait and the computer shuts itself off.

**Screenshots or videos showing encountered issue:** 
Impossible to get video as entire computer crashes

**osu!lazer version:** 
2020.312.0

**Logs:**
Unfortunately no relevant info found in osu log files, no signs of errors. Only this found in the system crash reports:
[WindowServer_2020-03-15-024502_Ethans-MacBook-Pro.log](https://github.com/ppy/osu/files/4334007/WindowServer_2020-03-15-024502_Ethans-MacBook-Pro.log)

**Computer Specifications:** 
MacBook Pro 2019 16 inch (MacBookPro 16,1)
macOS Catalina 10.15.3
2.6 GHz 6-Core Intel Core i7
AMD Radeon Pro 5300M
Display settings set to "More Space" (looks like 2048x1280)

**Comment by bdach** [comment]
Not a macOS owner but this looks pretty nasty and we might not be able to do anything to mitigate it:
```
Assertion failed: (isReady - AMD: AMD transaction hang or work overload (AMD IOFB)
DisplayID: 0x4281106
	Surface Use Counts: 70(1) 65(1) 1(3) 
	FB RegID: 4294969803, On Glass SurfaceIDs: 70 65 1, Transactions: [ Active: SurfaceID: 70 ] [ GPU Busy: SurfaceID: 65 ] [ GPU Busy: SurfaceID: 1 ]
	FB Driver Info: 0x4000 0x0 0x20 0x2 0x3 0x8 0xff0000 0xff00 0xff 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x2d2d2d2d 0x2d2d2d2d 0x52525252 0x52525252 0x47474747 0x47474747 0x42424242 0x42424242 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1000 0xa00 0x0 0x0
	3D Driver Info: 0x4000 0x0 0x20 0x2 0x3 0x8 0xff0000 0xff00 0xff 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x2d2d2d2d 0x2d2d2d2d 0x52525252 0x52525252 0x47474747 0x47474747 0x42424242 0x42424242 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x0 0x1000 0xa00 0x0 0x0
), function CoreDisplay_NotReady, file /BuildRoot/Library/Caches/com.apple.xbs/Sources/CoreDisplay/CoreDisplay-186.3.12/CoreDisplay/Display/Display.cpp, line 2957.
```
Honestly looks like a hardware/firmware issue.

**Comment by chaNcharge** [comment]
I'll update once I get the chance to test on the default display settings without night shift or any other display changes applied. Hopefully it's just a bug that reproduces in certain conditions so there's an easy temporary workaround.

**Comment by chaNcharge** [comment]
Unfortunately, default display settings without night shift activated does not fix any of the issues, so this could likely be a bug on Apple's side of things. I couldn't capture anything from the system logs or osu logs, as there appears to be a 4 minute gap in logging info in the system logs between normal operation and boot time, but if someone wants to try to reproduce this and be able to report on this, that'd be great.

**Comment by naoey** [comment]
I believe this might be an issue with the drivers/firmware in this particular model as I've also encountered this on my MBP 16,1 and the crash reports from those events also indicated something going wrong in the AMD driver.

**Comment by chaNcharge** [comment]
So, I believe the best thing to do right now is wait until Apple releases either a new version of macOS or a new beta that updates the drivers and test again?

**Comment by efarraro** [comment]
Not sure if this is at all relevant to this conversation as the use-case is quite different, but I stumbled across this issue while searching for the same bug. I also get this on MBP 16,1 with Unity (game engine), so it seems it may be a general issue with OS X.

**Comment by peppy** [comment]
@efarraro thanks for the added info! does unity still use GL?

**Comment by efarraro** [comment]
@peppy At least on OS X, I *think* it uses Metal now by default.

**Comment by cgkpdx** [comment]
FYI - this is happening to me using Unity in a large scene on a 2020 iMac with the high end Radeon card.

**Comment by peppy** [comment]
@cgkpdx thanks. is this issue being tracked anywhere else (ie. on the unity forums)? seems like this is only going to get resolved if someone reports to apple via a radar filing.

**Comment by cgkpdx** [comment]
I worked with an Apple senior tech the other day. Unfortunately I don't think this is going anywhere. The error seems to be a generic message for some sort of miscommunication between whatever app requesting GPU processing and the driver. It's not "specific enough" if that makes sense, and I don't think it's excessive to Unity. I did a bunch of digging around Google last night, and there are report of this issue with lots of different applications. What ended up solving it for me (so far) was to unplug my secondary display from my iMac. I "shouldn't have to" do that, but I'm just telling you what worked for me. Hardware diagnostic utility showed nothing wrong, safe mode was good, so at this point I'm seriously doubting a hardware issue. I mention that because it seems to be the first thing people on other forums jump to. Regarding tracking on the Unity forums...not really. I updated my Unity install to 2020.2.2f1 from the 2019 LTS version I had been using, partially because in the change log it notes an update to Metal API required by MacOS. I'm at a bit of a loss as to where to go from here. If you have a direction for me to go in that would be helpful for others just let me know and I'll see what I can do. 
[Windows Server Crash Report.txt](https://github.com/ppy/osu/files/5857468/Windows.Server.Crash.Report.txt)
